src/TransformationUtils.cc

```diff
@@ -88,57 +88,63 @@ std::unique_ptr<TransitionSystem> toTransitionSystem(ChcDirectedGraph const & gr
     return ts;
 }
 
-bool strongConnection(std::unordered_set<int> & visitedVertices, std::unordered_set<int> & verticesOnStack,
+// This function follows a Tarjan's strong connection detection algorithm:
+// https://en.wikipedia.org/wiki/Tarjan's_strongly_connected_components_algorithm
+// If some vertice is added on stack and it was already visited, then the strong connection is detected
+// It was updated to remember the nodes which participate in the loop (lines 123, 127)
+void strongConnection(std::unordered_set<SymRef, SymRefHash> & visitedVertices,
+                      std::unordered_set<SymRef, SymRefHash> & verticesOnStack,
                       AdjacencyListsGraphRepresentation const & graphRepresentation, ChcDirectedGraph const & graph,
-                      SymRef node) {
-    visitedVertices.insert(node.x);
-    verticesOnStack.insert(node.x);
+                      SymRef node, std::vector<EId> & loop) {
+    visitedVertices.insert(node);
+    verticesOnStack.insert(node);
     auto const & outEdges = graphRepresentation.getOutgoingEdgesFor(node);
     if (size(outEdges) <= 1) {
-        verticesOnStack.erase(node.x);
-        return false;
+        verticesOnStack.erase(node);
+        return;
     }
 
     for (EId eid : outEdges) {
         if (graph.getTarget(eid) != node) {
             auto nextVertex = graph.getTarget(eid);
-            if (visitedVertices.find(nextVertex.x) == visitedVertices.end()) {
-                bool loopFound =
-                    strongConnection(visitedVertices, verticesOnStack, graphRepresentation, graph, nextVertex);
-                if (loopFound) { return true; }
-            } else if (verticesOnStack.find(nextVertex.x) != verticesOnStack.end()) {
-                return true;
+            if (visitedVertices.find(nextVertex) == visitedVertices.end()) {
+                strongConnection(visitedVertices, verticesOnStack, graphRepresentation, graph, nextVertex, loop);
+                if (!loop.empty() && graph.getTarget(eid) != graph.getTarget(loop[0]) &&
+                    graph.getTarget(eid) == graph.getSource(loop[loop.size() - 1])) {
+                    loop.push_back(eid);
```

Comment:
Can you explain this condition?

Reply:
I check if the loop I found is actually loop:
graph.getTarget(eid) != graph.getTarget(loop[0]) &&
                    graph.getTarget(eid) == graph.getSource(loop[loop.size() - 1]
This checks if the target of the last edge of the loop and the source of the first edge is the same